Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean

```diff
@@ -0,0 +1,298 @@
+derivatives are independent of the representative for $[x:y:z]$, and the nonsingularity condition
+already implies that $(x, y, z) \ne (0, 0, 0)$, so a nonsingular rational point on `W` can simply be
+given by a tuple consisting of $[x:y:z]$ and the nonsingular condition on any representative.
+
+## Main definitions
+
+ * `WeierstrassCurve.Projective.PointClass`: the equivalence class of a point representative.
+ * `WeierstrassCurve.Projective.toAffine`: the Weierstrass curve in affine coordinates.
+ * `WeierstrassCurve.Projective.nonsingular`: the nonsingular condition on a point representative.
+ * `WeierstrassCurve.Projective.nonsingular_lift`: the nonsingular condition on a point class.
+
+## Main statements
+
+ * `WeierstrassCurve.Projective.polynomial_relation`: Euler's homogeneous function theorem.
+
+## Implementation notes
+
+A point representative is implemented as a term `P` of type `Fin 3 → R`, which allows for the vector
+notation `![x, y, z]`. However, `P` is not definitionally equivalent to the expanded vector
+`![P x, P y, P z]`, so the auxiliary lemma `fin3_def` can be used to convert between the two forms.
+The equivalence of two point representatives `P` and `Q` is implemented as an equivalence of orbits
+of the action of `Rˣ`, or equivalently that there is some unit `u` of `R` such that `P = u • Q`.
+However, `u • Q` is again not definitionally equal to `![u * Q x, u * Q y, u * Q z]`, so the
+auxiliary lemmas `smul_fin3` and `smul_fin3_ext` can be used to convert between the two forms.
+
+## References
+
+[J Silverman, *The Arithmetic of Elliptic Curves*][silverman2009]
+
+## Tags
+
+elliptic curve, rational point, projective coordinates
+-/
+
+local notation "x" => 0
+
+local notation "y" => 1
+
+local notation "z" => 2
+
+local macro "matrix_simp" : tactic =>
+  `(tactic| simp only [Matrix.head_cons, Matrix.tail_cons, Matrix.smul_empty, Matrix.smul_cons,
+              Matrix.cons_val_zero, Matrix.cons_val_one, Matrix.cons_val_two])
+
+universe u
+
+lemma fin3_def {R : Type u} (P : Fin 3 → R) : P = ![P x, P y, P z] := by
+  ext n; fin_cases n <;> rfl
+
+lemma smul_fin3 {R : Type u} [CommRing R] (P : Fin 3 → R) (u : Rˣ) :
+    u • P = ![u * P x, u * P y, u * P z] := by
+  rw [fin3_def P]
+  matrix_simp
+  rfl
+
+lemma smul_fin3_ext {R : Type u} [CommRing R] (P : Fin 3 → R) (u : Rˣ) :
+    (u • P) 0 = u * P x ∧ (u • P) 1 = u * P y ∧ (u • P) 2 = u * P z :=
+  ⟨rfl, rfl, rfl⟩
+
+/-! ## Weierstrass curves -/
+
+/-- An abbreviation for a Weierstrass curve in projective coordinates. -/
+abbrev WeierstrassCurve.Projective :=
+  WeierstrassCurve
+
+namespace WeierstrassCurve.Projective
+
+open MvPolynomial
+
+local macro "eval_simp" : tactic =>
+  `(tactic| simp only [eval_C, eval_X, eval_add, eval_sub, eval_mul, eval_pow])
+
+variable (R : Type u) [CommRing R]
+
+/-- The equivalence setoid for a point representative. -/
+def PointSetoid : Setoid <| Fin 3 → R :=
```

Comment:
Have you considered using https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Projectivization/Basic.html#Projectivization ? It could be applied to the vector space `Fin 3 -> R`.
It assumes that the base ring is a division ring, instead of a commutative ring. But that could probably be generalized.
Duplicating that construction will probably lead to a lot of API duplication down the road.

Reply:
Yes, that was the very first thing I considered, but there are a couple of annoyances at the top of my head:
- `Projectivization` has an extra layer of `v ≠ 0` for every vector `v` in `V`, which is not necessary here because it's already implied by `nonsingular`, and I prefer not to have that extra layer. I don't think anyone is currently developing API for concrete nonsingular varieties defined in terms of formal multivariate derivatives.
- As you mentioned, we need this for more than just division rings, and for an arbitrary commutative ring the correct definition is [slightly subtler](https://mathoverflow.net/questions/46116/a-valued-points-of-projective-space), as discussed in the elliptic curves topic in Zulip. We only need these for rings with trivial Picard group so I'm not sure how to incorporate this general restriction into the `Projectivization` file, so currently it's done ad-hoc.
- We don't have `WeightedProjectivization`, so the analogous construction would need to be made for `Jacobian`. If this is done down the line, and everything else works out, I'll be happy to switch over.

I imagine that there should not be too much API duplication, in the sense that the current duplicated code (99 -- 107 in this file) should be most of the duplicate API ever written for our purposes.